Chocotastic/CartViewController.swift

```diff
@@ -40,6 +44,34 @@ extension CartViewController {
     super.viewDidLoad()
     title = "Cart"
     configureFromCart()
+    setupCellConfiguration()
+    setupTotalCostConfiguration()
+  }
+}
+
+//MARK: - Rx Setup
+private extension CartViewController {
+  // ショッピングカート内のチョコレートをテーブルビューにバインド
+  func setupCellConfiguration() {
+    let cart = ShoppingCart.sharedCart
+    cart.chocolates
+      .map { Set($0) } // チョコレートの種類をユニーク化
+      .map { $0.sorted { $0.countryName > $1.countryName } } // セルの並びを国名でソート
+      .bind(to: tableView.rx.items(cellIdentifier: CartCell.Identifier, cellType: CartCell.self)) { index, chocolate, cell in
+        cell.configure(with: chocolate, cart: cart)
+      }
+      .disposed(by: disposeBag)
+  }
+  
+  // ショッピングカート内の合計値の変化を監視
+  func setupTotalCostConfiguration() {
+    let cart = ShoppingCart.sharedCart
+    cart.chocolates
+      .subscribe(onNext: { [unowned self] _ in
+        let cost = cart.totalCost
+        totalCostLabel.text = CurrencyFormatter.dollarsFormatter.string(from: cost)
```

Comment:
クロージャからのselfのプロパティを参照する場合、リークを防ぐために弱参照で参照します。

```suggestion
        self?.totalCostLabel.text = CurrencyFormatter.dollarsFormatter.string(from: cost)
```

Reply:
ただ、今回の実装では、Model側にCartが変わるたびにtotalCostをリアクティブに算出するBehaviorRelayをもたせて、Controllerでbindする方がMVVM的には自然かと思います。

```swift
class ShoppingCart {
  static let sharedCart = ShoppingCart()
  /// ショピングカート内のchocolatesを監視
  let chocolates: BehaviorRelay<[Chocolate]> = BehaviorRelay(value: [])
  let totalCost: BehaviorRelay<Float> = BehaviorRelay(value: 0.0)
  private let disposeBag = DisposeBag()

  func viewDidLoad() {
    bindingRx()
  }

  func bindingRx() {
    chocolates
      .map {
        $0.reduce(0.0) {
          runningTotal, chocolate in
          return runningTotal + chocolate.priceInDollars
        }
      }
      .bind(to: totalCost)
      .disposed(by: disposeBag)
  }
}
```


```swift
class ShoppingCart {
  static let sharedCart = ShoppingCart()
  /// ショピングカート内のchocolatesを監視
  let chocolates: BehaviorRelay<[Chocolate]> = BehaviorRelay(value: [])
  let totalCost: BehaviorRelay<Float> = BehaviorRelay(value: 0.0)
  private let disposeBag = DisposeBag()

  func viewDidLoad() {
    bindingRx()
  }

  func bindingRx() {
    chocolates
      .map {
        $0.reduce(0.0) {
          runningTotal, chocolate in
          return runningTotal + chocolate.priceInDollars
        }
      }
      .bind(to: totalCost)
      .disposed(by: disposeBag)
  }
}
```

CartViewController
```swfit
  func setupTotalCostConfiguration() {
    ShoppingCart.sharedCart
      .totalCost
      .distinctUntilChanged()
      .compactMap { CurrencyFormatter.dollarsFormatter.string(from: $0) }
      .bind(to: totalCostLabel.rx.text)
      .disposed(by: disposeBag)
  }
```